File: minStack.py
# MinStack() initializes the stack object.
# void push(int val) pushes the element val onto the stack.
# void pop() removes the element on the top of the stack.
# int top() gets the top element of the stack.
class MinStack:
    def minStack(self):
        global stack, min_stack
        stack = []
        min_stack = []
        print(stack)
    def push(self,a):
        if len(min_stack) == 0 or a < min_stack[-1]:
            min_stack.append(a)
        stack.append(a)
        print(stack)
    def pop(self):
        if stack[-1] == min_stack[-1]:
            min_stack.pop()
        stack.pop()
        print(stack)
    def top(self):
        if len(stack) > 0:
            print(stack[-1])
    def getMin(self):
        val = min_stack[-1]
        print(val)
    # Input
    # ["MinStack","push","push","push","getMin","pop","top","getMin"]
    # [[],[-2],[0],[-3],[],[],[],[]]
if __name__ == '__main__':
    s1 = MinStack()
    s1.minStack()
    s1.push(-2)
    s1.push(0)
    s1.push(-3)
    s1.getMin()
    s1.pop()
    s1.top()
    s1.getMin()

File: Jacob Gichira Mbuco/Stack & Queues/MinStack.java
import java.util.Stack;
/**
 Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.
	
	 
	
	Implement the MinStack class:
	
	 
	
	MinStack() initializes the stack object.
	void push(int val) pushes the element val onto the stack.
	void pop() removes the element on the top of the stack.
	int top() gets the top element of the stack.
	int getMin() retrieves the minimum element in the stack.
	 
	
	 
	
	Example:
	
	
	Input
	["MinStack","push","push","push","getMin","pop","top","getMin"]
	[[],[-2],[0],[-3],[],[],[],[]]
	
	 
	
	Output
	[null,null,null,null,-3,null,0,-2]                                                                                                                                                  
 */
//implement two stacks peek the min value of integers
public class MinStack {
    Stack<Integer> stack;
    Stack<Integer> minStack;
    /** initialize your data structure here. */
    public MinStack() {
        stack = new Stack<Integer>();
        minStack = new Stack<Integer>();
    }
    
    public void push(int x) {
        stack.push(x);
        if(minStack.isEmpty() || x <= minStack.peek()){
            minStack.push(x);
        }
    }
    
    public void pop() {
        if(stack.peek() == minStack.peek()){
            minStack.pop();
        }
        stack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minStack.peek();
    }
    //main method
    public static void main(String[] args) {
        MinStack minStack = new MinStack();
        minStack.push(-2);
        minStack.push(0);
        minStack.push(-1);
        System.out.println(minStack.getMin());
        minStack.pop();
        System.out.println(minStack.top());
        System.out.println(minStack.getMin());
    }

    

    
}
